topology/src/main/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImpl.java

```diff
@@ -220,4 +248,105 @@ private void failFuture(final ActorFuture<?> future, final Throwable error) {
           new TopologyRequestFailedException.InternalError(error.getMessage()));
     }
   }
+
+  /**
+   * Cancel the current operation. This is an inherently unsafe operation because there is no
+   * coordination among the nodes. To make it slightly safer, we first check if the node that is
+   * currently applying the operation is the same as that is in the local ClusterTopology. If it is
+   * not, then we have to get the latest topology from the other nodes before cancelling the change.
+   * This is still not safe if the changes are being applied while cancelling. We expect that cancel
+   * is invoked only when the operation is actually stuck and cannot make progress by itself.
+   *
+   * @param changeId id of the change to be cancelled
+   * @param currentClusterTopology current clusterTopology known to this node
+   * @param future will be completed when the change is cancelled or on failure
+   */
+  private void checkAndCancel(
```

Comment:
:question: Since cancelling an ongoing change is unsafe, I tried to add some checks to make it safer. But this also doesn't make it fully safe. So I'm also wondering if we should keep it simple, and just cancel it without these additional checks.

Reply:
I think I'd rather _not_ have these checks. Just as an example of how the checks could hurt us: if we had a bug where a topology update can't be merged, we'd need to cancel the change. But with these checks in place, they would sync and then try to merge and _also_ fail. If

Reply:
:+1: I will remove these checks. Will keep only the basic one such as local changeId is same as the requested one.